README.md
# Flask OAuth Filter - an OAuth Interceptor Logic

[![Quality](https://img.shields.io/badge/quality-production-green)](https://curity.io/resources/code-examples/status/)
[![Availability](https://img.shields.io/badge/availability-binary-blue)](https://curity.io/resources/code-examples/status/)

This library provides an extension for protecting APIs with OAuth when using Flask.

## Installation

You can either install this module with pip:
```pip install -U flask-of-oil```

Or copy the flask_of_oil folder in your project and install the requirements in ```requirements.txt``` using pip or pipenv

## Usage

This filter can be used in two ways with Flask, either to run before all routes with the same authorization requirement,
which protects all endpoints. Many times however it's desirable to only protect certain endpoints, which then can be done
using the `decorator` pattern described below.

### Using filter as a before_request filter that runs before ALL routes

When running the filter [before_request](http://flask.pocoo.org/docs/0.11/api/#flask.Flask.before_request)  *all* routes, the same configuration will apply to all routes. So if the filter is
 configured to require a scope of "read" then all routes will require that scope. If routes have different needs then
 the decorator pattern should be used (see next section).

**Example using before_request**

```python
import json
from flask import g, Flask
from flask_of_oil.oauth_filter import OAuthFilter

_app = Flask(__name__)
_oauth = OAuthFilter(verify_ssl=True)
_app.before_request(_oauth.filter)

@_app.route('/hello_world')
def hello_world():
    """
    :return: Returns a very useful JSON message when accessed.
    """
    print("OAuth Access token used for access")
    return json.dumps({"hello": g.user})
```


### Using filter as a decorator that runs before specific routes

Instead of setting the `before_request` a decorator can be added to the route that should be protected. This also enables the routes to have
 different scope requirements which could be handy.

*Important: The oauth decorator needs to be CLOSEST to the function*

```python
import json
from flask import g, Flask
from flask_of_oil.oauth_filter import OAuthFilter

_app = Flask(__name__)
_oauth = OAuthFilter(verify_ssl=True)

@_app.route('/hello_world')
@_oauth.protect(["read"])
def hello_world():
    """
    :return: Returns a very useful JSON message when accessed.
    """
    print("OAuth Access token used for access")
    return json.dumps({"hello": g.user})
```

### Authorizing the request based on scopes

The scope parameter of the protect decorator must either be a list or a space separated string:
``` 
["scope1", "scope2]
or 
"scope1 scope2"
```

### Authorizing the request based on claims

The incoming request can also be authorized based on claims, or a combination of claims and scopes.
The claims parameter of the protect decorator method has to be a `dict`, with keys the claims that are required 
for the request to be allowed. The value `None` instructs the filter to not check the value for the specific claim. 

```python
# Only allow requests where the incoming access token has the scope read and it contains a claim named MyGoodClaim
@_oauth.protect(scopes=["read"], claims={"MyGoodClaim": None})
```
```python
# Only allow requests where the incoming access token has the scope write and it contains a claim named MyGoodClaim with value MyGoodValue
@_oauth.protect(scopes=["write"], claims={"MyGoodClaim": "MyGoodValue"})
```


## Initializing the filter

**Filter global variable**

The OAuth filter should be setup the same way as Flask, a global reference and then initialized in main (or with the application)
The initialization depends on the type of tokens received. See the following examples.

```python
from flask import g, Flask
from flask_of_oil.oauth_filter import OAuthFilter

_app = Flask(__name__)
_oauth = OAuthFilter(verify_ssl=True)
```

**Using Opaque tokens**

When using Opaque tokens, the filter needs to resolve the reference by calling the introspection endpoint of the
OAuth server, this endpoint requires client credentials so the API needs to be a client of the OAuth server with the
permission to introspect tokens.

```python
if __name__ == '__main__':
    # configure the oauth filter
    _oauth.configure_with_opaque("https://oauth-server-host/oauth/v2/introspection", "api-client-id", "api-client-secret")

    # initiate the Flask app
    _app.run("localhost", debug=True, port=8000,
             ssl_context="adhoc")
```

**Using JWT tokens**

When using JWT (JWS) tokens, the filter will validate the signature of the token with the key that is provided on the
JWKS (Json Web Key Service) endpoint. The JWT contains a key id (kid) that is matched against the available public keys
on the OAuth server and then validated with that key.

```python
if __name__ == '__main__':
    # configure the oauth filter
    _oauth.configure_with_jwt("https://oauth-server-host/oauth/v2/metadata/jwks", "configured-issuer", "audience-of-token")

    # initiate the Flask app
    _app.run("localhost", debug=True, port=8000,
             ssl_context="adhoc")
```

**Using JWTs with multiple issuers**

You can configure multiple issuers of your tokens. The validator will match the issuer from the incoming JWT with one from
the configured list to properly validate the token.

```python
if __name__ == '__main__':
    # configure the oauth filter
    _oauth.configure_with_multiple_jwt_issuers(["https://idsvr.example.com/configured-issuer", "https://idsvr.example.com/another-issuer"], "audience-of-token")

    # initiate the Flask app
    _app.run("localhost", debug=True, port=8000,
             ssl_context="adhoc")
```

## Access token claims in Request object

When the filter accepts the request, it sets the `request.claims` context local variable for that request with all
the token claims. For JWT tokens, this is the JWT payload and for opaque tokens the introspection response. 

For example, the subject of the Authorization can be accessed like so `request.claims["sub"]`.

## Handling errors

The filter may abort the request if the Access token is invalid or if the scopes or claims in the access token doesn't match the
required scopes or claims for the route.

**401 Unauthorized**

When an invalid token is presented the filter will give a 401 unauthorized.
To customize the response, use Flasks [errorhandler](http://flask.pocoo.org/docs/0.11/api/#flask.Flask.errorhandler) to add a response.

```python
@_app.errorhandler(401)
def unauthorized(error):
    return json.dumps({'error': "unauthorized",
                       "error_description": "No valid access token found"}), \
           401, {'Content-Type': 'application/json; charset=utf-8'}
```

**403 Forbidden**

When a valid token is presented the filter but it's missing the appropriate scopes or claims (or the claims have wrong values)
then the request is aborted with a 403 Forbidden.

```python
@_app.errorhandler(403)
def forbidden(error):
    return json.dumps({'error': "forbidden",
                       "error_description": "Access token is missing appropriate scopes"}), \
           403, {'Content-Type': 'application/json; charset=utf-8'}
```

## Running the test suite

You can run the test suite with the following command:

```bash
pytest --log-level=DEBUG
```

If you want to run tests from a concrete file, use this command:

```bash
pytest --log-level=DEBUG tests/test_jwt_validator.py
```

It's best to run the tests in a dedicated Python virtual environment. Follow these steps to create the environment, install all dependencies, then run the test suite:

```bash
python -m venv venv
source venv/bin/activate
pip install -r requirements.txt
pytest --log-level=DEBUG tests/test_jwt_validator.py
```

When you're done with the tests, run:

```bash
deactivate
```

to exit the virtual environment.

## Questions and Support

For questions and support, contact Curity AB:

> Curity AB
>
> info@curity.io
> https://curity.io


Copyright (C) 2016 Curity AB.


CHANGELOG.md
## Flask-of-oil changelog

## Unreleased

## 1.2.2 (2023-05-29)

Feature:
    - Updated requests to version 2.31.0

## 1.2.1 (2023-05-29)

Feature:
    - Updated flask to version 2.3.2.

Fix:
    - Properly set empty scope in JWT payload, if no scope present in the JWT.

## 1.2.0 (2022-05-12)

Feature:
    - Refresh JWKS from JWKS URI when no suitable keys found in current JWKS.

Fix:
    - Fixed verification of `aud` claim. Now properly verifies audience where a token contains a single audience value.

## 1.1.0 (2021-05-05)

Feature:
    - Allow configuring multiple issuers.

Fix:
    - Return default errors in all cases so that they can be properly caught by flask error handlers.
    - You can now omit passing required claims.

## 1.0.1 (2020-02-21)

Initial release.


flask_of_oil/oauth_filter.py
# #########################################################################
# Copyright 2016 Curity AB
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

import logging
import json

from flask import request, abort
from flask_of_oil.jwt_validator import JwtValidator
from flask_of_oil.opaque_validator import OpaqueValidator
from functools import wraps

from flask_of_oil.tools import base64_urldecode


class OAuthFilter:
    def __init__(self, verify_ssl=True):
        self.protected_endpoints = {}
        self.configured = False
        self.verify_ssl = verify_ssl
        self.logger = logging.getLogger(__name__)
        self.validators = dict()
        self.scopes = list()

    def configure_with_jwt(self, jwks_url, issuer, audience, scopes=None):
        """

        :param jwks_url:
        :param issuer:
        :param audience:
        :param scopes:
        :return:
        """
        self.validators["*"] = JwtValidator(jwks_url, issuer, audience, self.verify_ssl)

        if scopes is not None:
            self.scopes = scopes

    def configure_with_multiple_jwt_issuers(self, issuers, audience, scopes=None):
        """

        :param issuers: List of issuer values.
        :param audience: The expected value of the aud claim.
        :param scopes: List of required scopes.
        :return:
        """

        if scopes is not None:
            self.scopes = scopes

        for issuer in issuers:
            jwks_url = issuer + "/jwks"
            self.validators[issuer] = JwtValidator(jwks_url, issuer, audience, self.verify_ssl)

    def configure_with_opaque(self, introspection_url, client_id, client_secret, scopes=None):
        """

        :param introspection_url:
        :param client_id:
        :param client_secret:
        :param scopes:
        :return:
        """
        self.validators["*"] = OpaqueValidator(introspection_url, client_id, client_secret, self.verify_ssl)

        if scopes is not None:
            self.scopes = scopes

    def _add_protected_endpoint(self, func, scopes):
        self.protected_endpoints[func] = scopes

    @staticmethod
    def _extract_access_token(incoming_request=None):
        """
        Extract the token from the Authorization header
        OAuth Access Tokens are placed in the header in the form "Bearer XYZ", so Bearer
        needs to be removed and the whitespaces trimmed.

        The method will abort if no token is present, and return a 401
        :param incoming_request: The incoming flask request
        :return: the stripped token
        """

        authorization_header = incoming_request.headers.get("authorization", type=str)
        query_param_access_token = incoming_request.args.get("access_token", type=str)

        if authorization_header is None and query_param_access_token is None:
            abort(401)

        if authorization_header is not None:
            authorization_header_parts = authorization_header.split()
            authorization_type = authorization_header_parts[0].lower()

            # Extract the token from the Bearer string
            if authorization_type != "bearer":
                abort(401)

            return authorization_header_parts[1] if len(authorization_header_parts) >= 2 else None

        return query_param_access_token

    def _authorize(self, token_claims, endpoint_scopes, endpoint_claims):

        if endpoint_claims is None:
            endpoint_claims = {}

        for claim in endpoint_claims:
            if claim not in token_claims or \
                    (endpoint_claims[claim] is not None and token_claims[claim] != endpoint_claims[claim]):
                return False

        scope = token_claims['scope']
        if isinstance(token_claims['scope'], (list, tuple)):
            incoming_scopes = scope
        else:
            incoming_scopes = scope.split()

        if endpoint_scopes is None:
            required_scopes = self.scopes
        else:
            required_scopes = endpoint_scopes

        return all(s in incoming_scopes for s in required_scopes)

    def protect(self, scopes=None, claims=None):
        """
        This is a decorator function that can be used on a flask route:
        @_oauth.protect(["read","write]) or @_oauth.protect()
        :param claims: The claims that are required for the protected endpoint (dict)
        :param scopes: The scopes that are required for the protected endpoint (list or space separated string)
        """

        if scopes is None:
            scopes = []

        if not isinstance(scopes, list):
            scopes = scopes.split()

        if claims is None:
            claims = {}

        if not isinstance(claims, dict):
            claims = {}
            self.logger.warning("claims is not a dict and will be ignored")

        def decorator(f):
            @wraps(f)
            def inner_decorator(*args, **kwargs):
                if self.filter(scopes=scopes, claims=claims) is None:
                    return f(*args, **kwargs)
                else:
                    abort(500)

            return inner_decorator

        return decorator

    def filter(self, scopes=None, claims=None):
        self.logger.debug("Request method = " + str(request.method))
        token = self._extract_access_token(request)
        validated_token = dict(active=False)
        self.logger.debug("Access token " + token)

        # noinspection PyBroadException
        try:
            if "*" in self.validators:
                validated_token = self.validators["*"].validate(token)
            else:
                # See if we can determine the issuer; we'll use that as the key to find
                # the validator.
                parts = token.split(".")

                if len(parts) == 3:
                    payload = json.loads(base64_urldecode(parts[1]))
                    issuer = payload["iss"]

                    if issuer in self.validators:
                        validated_token = self.validators[issuer].validate(token)
        except Exception:
            abort(500)
            return

        if not validated_token['active']:
            abort(401)

        # Authorize scope
        authorized = self._authorize(validated_token, endpoint_scopes=scopes, endpoint_claims=claims)
        if not authorized:
            abort(403)

        # Set the user info in a context global variable
        request.claims = validated_token

        return None
